src/app/tests/TestThreadBorderRouterManagementCluster.cpp

```diff
@@ -0,0 +1,334 @@
+        mActiveDatasetLen                   = activeDataset.AsByteSpan().size();
+        mCallback                           = callback;
+        mSetActiveDatasetCommandSequenceNum = sequenceNumber;
+    }
+
+    CHIP_ERROR CommitActiveDataset() override { return CHIP_NO_ERROR; }
+
+    CHIP_ERROR RevertActiveDataset() override
+    {
+        mStoredActiveDatasetLen = 0;
+        mInterfaceEnabled       = false;
+        mCallback               = nullptr;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR SetPendingDataset(const Thread::OperationalDataset & pendingDataset) override
+    {
+        memcpy(mPendingDataset, pendingDataset.AsByteSpan().data(), pendingDataset.AsByteSpan().size());
+        mPendingDatasetLen = pendingDataset.AsByteSpan().size();
+        return CHIP_NO_ERROR;
+    }
+
+    void ActivateActiveDataset()
+    {
+        memcpy(mStoredActiveDataset, mActiveDataset, Thread::kSizeOperationalDataset);
+        mStoredActiveDatasetLen = mActiveDatasetLen;
+        mInterfaceEnabled       = true;
+        if (mCallback)
+        {
+            mCallback->OnActivateDatasetComplete(mSetActiveDatasetCommandSequenceNum, CHIP_NO_ERROR);
+        }
+        mCallback = nullptr;
+    }
+
+    bool mPanChangeSupported  = true;
+    const char * kTestName[2] = { "TestName", "TestNameLength64________________________________________________" };
+    const uint8_t kTestBorderAgentId[kBorderAgentIdLength]        = { 0x0,  0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77,
+                                                                      0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF };
+    const uint16_t kTestThreadVersion                             = 4;
+    uint8_t mActiveDataset[Thread::kSizeOperationalDataset]       = { 0 };
+    size_t mActiveDatasetLen                                      = 0;
+    uint8_t mStoredActiveDataset[Thread::kSizeOperationalDataset] = { 0 };
+    size_t mStoredActiveDatasetLen                                = 0;
+    uint8_t mPendingDataset[Thread::kSizeOperationalDataset]      = { 0 };
+    size_t mPendingDatasetLen                                     = 0;
+    bool mUseInvalidBorderRouterName                              = true;
+    size_t mTestBorderAgentIdLen                                  = kBorderAgentIdLength - 1;
+    bool mInterfaceEnabled                                        = false;
+    uint32_t mSetActiveDatasetCommandSequenceNum                  = 0;
+    ActivateDatasetCallback * mCallback                           = nullptr;
+};
+
+constexpr EndpointId kTestEndpointId            = 1;
+constexpr FabricIndex kTestAccessingFabricIndex = 1;
+static FailSafeContext sTestFailsafeContext;
+static TestDelegate sTestDelegate;
+static ServerInstance sTestSeverInstance(kTestEndpointId, &sTestDelegate, sTestFailsafeContext);
+
+class TestSetActiveDatasetCommandHandler : public CommandHandler
```

Comment:
Overall, if the logic of the cluster had been abstracted to a logic-only class called by the ServerInstance that implements CommandHandler and AttributeAccessInterface, none of this faking would be needed and it would be simpler to test the cluster.

Reply:
I am not clear on this. Do you mean that we should add a test function in the ServerInstance and run the test logic in that function?